### PR TITLE
envtest: add DlbDevicePlugin tests

### PR DESCRIPTION
**Pull request by hj-johannes-lee**
envtest:
- add DlbDevicePlugin tests

### CODE REVIEW

**Comment by mythi** [comment]
> out that there was an update of tutorial that we followed for suite_test.go file

Let's have these changes in a commit of it's own and update `ci.yaml` and `Jenkinsfile` part of it. Everything DLB related should go into its own commit.

**Review by mythi — CHANGES_REQUESTED** [review]
@hj-johannes-lee nice, thanks!

Let's just re-organize the changes.

**Comment by hj-johannes-lee** [comment]
@mythi Umm,, do you mean to change the previous commit (that is merged earlier) to include the changes of this pr?

**Comment by mythi** [comment]
@hj-johannes-lee that's not possible. I meant that just split the changes in this PR into the two commits differently.

**Review by mythi — DISMISSED** [review]
thanks!

File: test/envtest/dlbdeviceplugin_controller_test.go [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+				Image: "testimage",
+			}
+
+			key := types.NamespacedName{
+				Name: "dlbdeviceplugin-test",
+			}
+
+			toCreate := &devicepluginv1.DlbDevicePlugin{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: key.Name,
```
Comment (bart0sh) [inline test/envtest/dlbdeviceplugin_controller_test.go:51]: Why do we set plugin name to the namespace name?

File: test/envtest/dlbdeviceplugin_controller_test.go [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+			By("updating image name successfully")
+			updatedImage := "updated-dlb-testimage"
+			fetched.Spec.Image = updatedImage
+
+			Expect(k8sClient.Update(context.Background(), fetched)).Should(Succeed())
+			fetchedUpdated := &devicepluginv1.DlbDevicePlugin{}
+			Eventually(func() string {
+				_ = k8sClient.Get(context.Background(), key, fetchedUpdated)
+				return fetchedUpdated.Spec.Image
+			}, timeout, interval).Should(Equal(updatedImage))
```
Comment (bart0sh) [inline test/envtest/dlbdeviceplugin_controller_test.go:75]: Should we also check Args and Node selector updates?

File: test/envtest/dlbdeviceplugin_controller_test.go [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+		It("should handle DlbDevicePlugin objects correctly", func() {
+			spec := devicepluginv1.DlbDevicePluginSpec{
+				Image: "testimage",
+			}
+
+			key := types.NamespacedName{
+				Name: "dlbdeviceplugin-test",
+			}
+
+			toCreate := &devicepluginv1.DlbDevicePlugin{
```
Comment (bart0sh) [inline test/envtest/dlbdeviceplugin_controller_test.go:49]: what about to set namespace, Image and Args for container to cover this code? https://github.com/intel/intel-device-plugins-for-kubernetes/blob/main/pkg/controllers/dlb/controller.go#L78:L81

File: test/envtest/dlbdeviceplugin_controller_test.go [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+				ObjectMeta: metav1.ObjectMeta{
+					Name: key.Name,
+				},
+				Spec: spec,
+			}
+
```
Comment (bart0sh) [inline test/envtest/dlbdeviceplugin_controller_test.go:55]: I've noticed that NodeSelector updates are processed by the [UpdateDaemonSet](https://github.com/intel/intel-device-plugins-for-kubernetes/blob/main/pkg/controllers/dlb/controller.go#L94), but not processed by the [NewDaemonSet](https://github.com/intel/intel-device-plugins-for-kubernetes/blob/main/pkg/controllers/dlb/controller.go#L74). However, for other plugins this doesn't seem to be the case, e.g. GPU plugin takes care of NodeSelector in both places: [NewDaemonSet](https://github.com/intel/intel-device-plugins-for-kubernetes/blob/main/pkg/controllers/gpu/controller.go#L123:L134) and [UpdateDaemonSet](https://github.com/intel/intel-device-plugins-for-kubernetes/blob/main/pkg/controllers/gpu/controller.go#L247:L255)

Comment (hj-johannes-lee) [reply test/envtest/dlbdeviceplugin_controller_test.go:55]: I think I solved the problem in #764!

File: test/envtest/dlbdeviceplugin_controller_test.go (hunk 1 above)
Comment (mythi) [reply test/envtest/dlbdeviceplugin_controller_test.go:51]: The CRDs are cluster scoped: see 37618d4f850f63ae0503b0b313242540a9bb269d

Comment (bart0sh) [reply test/envtest/dlbdeviceplugin_controller_test.go:51]: not sure I follow. I was proposing to test [this line of code](https://github.com/intel/intel-device-plugins-for-kubernetes/blob/main/pkg/controllers/dlb/controller.go#L78)

Comment (mythi) [reply test/envtest/dlbdeviceplugin_controller_test.go:51]: the namespace where the daemonset is created is not taken from the CR namepace (they don't have one) so it's not possible to test that, at least with envtest

Comment (bart0sh) [reply test/envtest/dlbdeviceplugin_controller_test.go:51]: got it. thanks for the explanations!

File: test/envtest/dlbdeviceplugin_controller_test.go (hunk 2 above)
Comment (hj-johannes-lee) [reply test/envtest/dlbdeviceplugin_controller_test.go]: If I understood correctly, I fixed as you told!

File: test/envtest/dlbdeviceplugin_controller_test.go [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+			By("creating DlbDevicePlugin successfully")
+			Expect(k8sClient.Create(context.Background(), toCreate)).Should(Succeed())
+			time.Sleep(time.Second * 5)
+
+			fetched := &devicepluginv1.DlbDevicePlugin{}
+			Eventually(func() bool {
+				_ = k8sClient.Get(context.Background(), key, fetched)
+				return len(fetched.Status.ControlledDaemonSet.UID) > 0
+			}, timeout, interval).Should(BeTrue())
```
Comment (bart0sh) [inline test/envtest/dlbdeviceplugin_controller_test.go:66]: Can we check here if Image and other plugin attributes have expected values?

File: test/envtest/dlbdeviceplugin_controller_test.go [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+			Expect(k8sClient.Update(context.Background(), fetched)).Should(Succeed())
+			fetchedUpdated := &devicepluginv1.DlbDevicePlugin{}
+			Eventually(func() devicepluginv1.DlbDevicePluginSpec {
+				_ = k8sClient.Get(context.Background(), key, fetchedUpdated)
+				return fetchedUpdated.Spec
+			}, timeout, interval).Should(Equal(
+				devicepluginv1.DlbDevicePluginSpec{
+					Image:        updatedImage,
+					NodeSelector: updatedNodeSelector,
+					LogLevel:     updatedLogLevel}))
```
Comment (hj-johannes-lee) [inline test/envtest/dlbdeviceplugin_controller_test.go:85]: @bart0sh
> Can we check here if Image and other plugin attributes have expected values?

Umm, did I understand your previous comment wrongly? Is it what you meant..?
And, what do you mean "here"? You want me to move the code to there? Or, do you want me to use Should(BeTrue()) than SHould(Equal())?

Comment (bart0sh) [reply test/envtest/dlbdeviceplugin_controller_test.go:85]: > And, what do you mean "here"?

I meant after line 66, where I added that comment. No, I don't want you to move this code there. I want you to test attributes for both operations: Create(which you didn't do yet) and Update(which you've already done).

Comment (bart0sh) [reply test/envtest/dlbdeviceplugin_controller_test.go:85]: A little bit more details on this:
Lines 83-85 verify attributes (Image, Nodeselector, LogLevel) set by UpdateDaemonSet
However, changes made by NewDaemonSet (Args and Image) are not verified.

File: .github/workflows/ci.yaml [hunk 7]
```diff
@@ -11,7 +11,7 @@ on:
 env:
```
Comment (bart0sh) [inline .github/workflows/ci.yaml:11]: @hj-johannes-lee This commit doesn't seem to depend on another one. I'd propose to submit them in a separate PRs.

File: test/envtest/dlbdeviceplugin_controller_test.go (hunk 6 above)
Comment (hj-johannes-lee) [reply test/envtest/dlbdeviceplugin_controller_test.go:85]: Ah, now I understand. I don't know why I could not see that there is no code for Create.!
I think I was confused since I didn't catch that you meant after line 66.!

Thank you for the detailed explanation!

File: test/envtest/dlbdeviceplugin_controller_test.go [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+	AfterEach(func() {
+		time.Sleep(time.Second * 2)
+
+	})
+
+	Context("Basic CRUD operations", func() {
+		It("should handle DlbDevicePlugin objects correctly", func() {
+			spec := devicepluginv1.DlbDevicePluginSpec{
+				Image:        "testimage",
+				NodeSelector: map[string]string{"feature.node.kubernetes.io/dlb": "true"},
```
Comment (bart0sh) [inline test/envtest/dlbdeviceplugin_controller_test.go:43]: we probably don't need to set NodeSelector here as it will be part of plugin yaml when your other PR is merged.

Comment (hj-johannes-lee) [reply test/envtest/dlbdeviceplugin_controller_test.go:43]: Umm,,, sorry, I don't understand two things.

1. Isn't what I did for [this matter](https://github.com/intel/intel-device-plugins-for-kubernetes/pull/760#discussion_r755996314)? I thought I understood what you meant but now I am confused.!

2. How is this related to another PR I made..? I thought this is more related to deployment/operator/samples/yaml file than the plugin yaml file and the operator?

Comment (bart0sh) [reply test/envtest/dlbdeviceplugin_controller_test.go:43]: 1. your link points to the top of this PR, so I didn't get what you mean
2. We need help from Mikko on this I guess. I thought we're testing controller code in envtests.

Comment (hj-johannes-lee) [reply test/envtest/dlbdeviceplugin_controller_test.go:43]: 1. Sorry. I didn't know the link does not go to there if it's resolved and closed. What I wanted to ask is "Was what I put for satisfying what you suggested?"

2. I see.! @mythi :) Please help us (maybe on Monday)!

Comment (mythi) [reply test/envtest/dlbdeviceplugin_controller_test.go:43]: I believe @hj-johannes-lee refers to [this comment](https://github.com/intel/intel-device-plugins-for-kubernetes/pull/760/files#r755996314).

None of our `envtest` cases check the resulting daemonset is as expected. We should start adding those checks we don't need them in this PR for DLB. This PR is OK as it is now.

Comment (hj-johannes-lee) [reply test/envtest/dlbdeviceplugin_controller_test.go:43]: @mythi
yeap! it is what I wanted to refer.!
Then, can I regard this comment as resolved without changing anything?

**Review by mythi — APPROVED** [review]

**Review by bart0sh — APPROVED** [review]
/lgtm
